merge.go

```diff
@@ -114,20 +119,41 @@ next:
 	return out
 }
 
-// lookupByHeader optionally returns a File from fs.files if the FileHeaders match.
+// findOutfile optionally returns a File from fs.files if the FileHeaders match.
 // This is done because we append batches into files to minimize the count of output files.
 //
-// lookupByHeader will return the existing file (stored in outfiles) if no matching file exists.
-func (fs *mergableFiles) lookupByHeader(f *File) *File {
-	for i := range fs.outfiles {
-		if fs.outfiles[i].Header.ImmediateDestination == f.Header.ImmediateDestination &&
-			fs.outfiles[i].Header.ImmediateOrigin == f.Header.ImmediateOrigin {
-			// found a matching file, so return it
-			return fs.outfiles[i]
+// findOutfile will return the existing file (stored in outfiles) if no matching file exists.
+func (fs *mergableFiles) findOutfile(f *File) *File {
+	var lookup func(int) *File
+	lookup = func(start int) *File {
+		// To allow recursive lookups we need to memorize the current index so deeper calls
+		// will bypass files with conflicting trace numbers.
+		for i := start; i < len(fs.outfiles); i++ {
+			if fs.outfiles[i].Header.ImmediateDestination == f.Header.ImmediateDestination &&
+				fs.outfiles[i].Header.ImmediateOrigin == f.Header.ImmediateOrigin {
+
+				// found a matching file, so verify the TraceNumber isn't alreay inside
+				outTraceNumbers := getTraceNumbers(fs.outfiles[i])
+				inTraceNumbers := getTraceNumbers(f)
+
+				for j := range inTraceNumbers {
+					// If any of our incoming trace numbers match the existing merged file
+					// return the entire file as separate. This keeps partially overlapping
+					// batches self-contained.
+					if outTraceNumbers.contains(inTraceNumbers[j]) {
+						return lookup(i + 1)
```

Comment:
Pretty interesting recursion going on here.
So if it gets to this point; we check out the next file in `fs.outfiles` and we either: use `fs.outfiles[i+1]` or create a new one

Reply:
We would check "i+1" through the end of the slice for another file. This is to accumulate into a second file when collisions occur.